### DISCUSSION TITLE
Remote control interface for Freeplane

### DISCUSSION THREAD

**Discussion by metacoma** [freeplane/freeplane#792] [Ideas]
Hello, 

Is there a way to send groovy commands (groovy-script) externally like
```bash
$ echo "node.createChild('foobar')" | unixcat /path/to/freeplane/control/unixsocket
$ cat << EOF | nc 127.0.0.1 1123
node['script1'] = "println 'Hello world'"
EOF
```
As i can understand, at this moment there is no way to do that.
My question, what the best way to implement this? Is it possible to make freeplane to listen unix/tcp/udp socket (in thread or async)? for example https://github.com/gesellix/unix-socket-factory/blob/master/src/test/groovy/socketfactory/unix/UnixSocketTestServer.groovy  

Thank you

**Comment by metacoma** [comment]
Looks like the batch job this is what i need https://docs.freeplane.org/attic/old-mediawiki-content/Batch_Jobs.html

```bash
$ ./freeplane.sh &
$ echo 'node.createChild("hello world")' > ~/.config/freeplane/*/scripts/test.groovy
$ bash ./freeplane.sh -SNXtest
```
But unfortunately it's not working in this way :(

**Comment by metacoma** [comment]
Hey, i will bump this thread.

I have a good explanation for my question and example how it will work.
My question is: is there cheapest/simple way to implement "remote control" in freeplane.
Under "remote control" i mean the "algorithm", for example is groovy freeplane script, or java freeplane plugin, that listen (read) groovy-code from TCP/IP socket, or UDP datagram, or read groovy-code from local fifo/unixsocket and then execute this groovy code in current mindmap context.

real-world usage example, i have a simple `users.yaml`:
```yaml
users:
  bob:
      uid: 1008
  alice:
      uid: 420
```
then i want execute following commands: (`yaml2freeplane-groovy.sh` is simple script that accept yaml in stdin and produce groovy code to stdout)

```bash
# example with unixsocket
$ cat users.yaml | yaml2freplane-groovy.sh >> /tmp/freeplane_socket
# read answer from freeplane
$ cat /tmp/freeplane_socket
  
# the same example with TCP/IP
$ cat users.yaml | yaml2freeplane-groovy.sh | nc localhost 3000   
```

After i ran these command, the current mindmap should change in this way:
![screenshot](https://user-images.githubusercontent.com/5146707/203422054-ef177b54-9965-4c5c-a835-50c22c97c423.png)

How i can achieve this in the easiest way?

Thank you.

**Comment by metacoma** [comment]
After some research, i figured out how to implement the simple remote control script.
Of course, this script need some polish, but basically it's works.

First of all, create fifo socket
```bash
$ mkfifo /tmp/fifo
``` 
Then apply following script on random freeplane node

```groovy
import java.util.Arrays;
import java.io.FileInputStream;
import java.io.ByteArrayOutputStream;

/******************************************************************/
def threads = new Thread[Thread.activeCount()]
Thread.currentThread().getThreadGroup().enumerate(threads)
Arrays.stream(threads).forEach(Thread::interrupt)

/******************************************************************/
def name = "fifo-thread"
new Thread(() -> {
   def pipe = new File('/tmp/fifo')
   def bos = new ByteArrayOutputStream()
   def len = -1
   byte[] buff = new byte[256]
   def i = 0
    while(1) {
        try {
            Thread.sleep(1000)
            def fis = new FileInputStream(pipe).eachLine { line ->
              Eval.me(line)
              println line
            }
        } catch (InterruptedException e) {
            break
        }
    }
},name).start()

```

And finally, send the "remote" command from the linux shell to freeplane instance through the /tmp/fifo file

```bash
$ echo 'println "hello world"' >> /tmp/fifo
```

After it, you should see in freeplane logs smth like this:
```
...
hello world
println "hello world"
...
```

**Reply by dpolivaev** [reply]
Because Freeplane is not thread-safe you can use

````
EventQueue.invokeAndWait(() -> Eval.me(line))
````

**Reply by metacoma** [reply]
> Because Freeplane is not thread-safe you can use
> 
> ```
> EventQueue.invokeAndWait(() -> Eval.me(line))
> ```

thank you for your suggestion, it's works like a charm
I'm not very familiar with groovy, could you please explain, how I can access to the node ("node" variable, for example the current selected node" context and freeplane context ("c" variable) from the separated java thread? 

for example, i want do smth like this
```bash
echo 'node.createChild("hello world")' >> /tmp/pipe 
```
but unfortunately, got java execption that the node variable is not defined in context where i try to evaluate the code.
```
....
Caused by: groovy.lang.MissingPropertyException: No such property: node for class: Script1
        at org.codehaus.groovy.runtime.ScriptBytecodeAdapter.unwrap(ScriptBytecodeAdapter.java:67)
        at org.codehaus.groovy.vmplugin.v8.IndyGuardsFiltersAndSignatures.unwrap(IndyGuardsFiltersAndSignatures.java:172)
        at org.codehaus.groovy.vmplugin.v8.IndyInterface.fromCache(IndyInterface.java:318)
        at Script1.run(Script1.groovy:1)
        at groovy.lang.GroovyShell.evaluate(GroovyShell.java:432)
        at groovy.lang.GroovyShell.evaluate(GroovyShell.java:466)
        at groovy.lang.GroovyShell.evaluate(GroovyShell.java:441)
        at groovy.util.Eval.me(Eval.java:67)
        at groovy.util.Eval.me(Eval.java:50)
...
```


Just in case, the "pipe" groovy-script looks as follows:
```groovy
import java.util.Arrays;
import java.io.FileInputStream;
import java.io.ByteArrayOutputStream;
import java.awt.EventQueue;

/******************************************************************/
def threads = new Thread[Thread.activeCount()]
Thread.currentThread().getThreadGroup().enumerate(threads)
Arrays.stream(threads).forEach(Thread::interrupt)

/******************************************************************/
def name = "fifo-thread"
new Thread(() -> {
   def pipe = new File('/tmp/pipe')
   def bos = new ByteArrayOutputStream()
   def len = -1
   byte[] buff = new byte[256]
   def i = 0
    while(1) {
        try {
            Thread.sleep(1000)
            def fis = new FileInputStream(pipe).eachLine { line ->
              // Eval.me(line)
              println "test"
              EventQueue.invokeAndWait(() -> Eval.me(line))
              println line
            }
        } catch (InterruptedException e) {
            break
        }
    }
},name).start()
```

Looks like, i need to somehow pass some context variables into new java-thread?
Thank you.

**Reply by metacoma** [reply]
> Looks like, i need to somehow pass some context variables into new java-thread? Thank you.

I found  some "workaround"
```bash
$ cat<< EOF >> /tmp/pipe
> import org.freeplane.plugin.script.proxy.ScriptUtils; def c = ScriptUtils.c(); c.statusInfo = 'hello world';
> EOF
```

It is works, but looks like i need to change my code in this way:
```groovy
def groovy_body = read_all_data(pipe)
EventQueue.invokeAndWait(() -> Eval.me("import org.freeplane.plugin.script.proxy.ScriptUtils; def c = ScriptUtils.c();\n"  + groovy_body))
```
Yes, it's a little bit ugly, but i think it should work

**Comment by metacoma** [comment]
the second version of groovy-fifo-remote-control script

before apply groovy-script on the some node, don't forget, run once in bash terminal:
```bash
$ mkfifo /tmp/fifo
```
`groovy-fifo-remote-control.groovy` content:
```groovy
import java.util.Arrays;
import java.io.FileInputStream;
import java.io.ByteArrayOutputStream;
import java.awt.EventQueue;
import org.freeplane.plugin.script.proxy.ScriptUtils

/******************************************************************/
def threads = new Thread[Thread.activeCount()]
Thread.currentThread().getThreadGroup().enumerate(threads)
Arrays.stream(threads).forEach(Thread::interrupt)

/******************************************************************/
def name = "fifo-thread"
new Thread(() -> {
  InputStream Fifo = new FileInputStream("/tmp/fifo");
  int buffer_size = 65536;
  byte[] bytes = new byte[0x10000];
  int bytes_read = -1;
  while(1) {
    while ((bytes_read = Fifo.read(bytes, 0, buffer_size)) == -1) {
//          print "."
      Thread.sleep(1000)
    }
    String text = new String(bytes, 0, bytes_read)
    String groovy_script = """
import org.freeplane.plugin.script.proxy.ScriptUtils;
def c = ScriptUtils.c();
def node = ScriptUtils.node();
""" + text;
    println groovy_script
    EventQueue.invokeAndWait(() -> Eval.me(groovy_script))
  }
},name).start()

```

works as expected

```bash
$ echo 'node.createChild("hello github")' >> /tmp/fifo
```
![screenshot](https://user-images.githubusercontent.com/5146707/204106448-d3c7d60b-efb6-4760-a005-323be199a7ea.png)

**Comment by metacoma** [comment]
And here is a result

https://user-images.githubusercontent.com/5146707/204116377-e34be6ed-2743-4c94-b8ac-c417ee2b8e12.mp4

**Reply by dpolivaev** [reply]
Looks impressive! Would you like to share your code?

**Reply by metacoma** [reply]
> Looks impressive! Would you like to share your code?

Hey, It will be great, if anyone, other than me will try this!

I have created the repository on github with all necessary scripts and step-by-step instructions how you can reproduce mindmap-shell workflow.

Please take a look https://github.com/metacoma/mindwm 

If you are faced with some issues, in this case, please, don't hesitate and provide the feedback in any way convenient for you.

**Reply by EdoFro** [reply]
Thanks!

**Comment by nnako** [comment]
Hi @metacoma . What a great demonstration of your concept for a "Freeplane remote control". Looks very interesting. I created a [question](https://github.com/metacoma/mindwm/issues/23) on your project page to get started... Thanks.

**Reply by metacoma** [reply]
Hey @nnako thank you!
I have attempted  to [answer](https://github.com/metacoma/mindwm/issues/23#issuecomment-1382720804) your questions thoroughly. Please feel free to ask anything.

**Comment by macmarrum** [comment]
@metacoma This is a great idea. I finally got around to playing with it.

I've created an `init` script which runs a Groovy-scripting server on localhost:48112, i.e. without the need for a `fifo`. This way, remote script execution can be enjoyed also on Windows.

Here's a simple client (groovy_scripting_client.py) written in Python.

```python
#!/usr/bin/python3
import sys
import socket

port = 48112

if len(sys.argv) == 2:
    text = sys.argv[1]
else:
    text = sys.stdin.read()

UTF8 = 'UTF-8'
s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
s.connect(('127.0.0.1', port))
s.sendall(text.encode(UTF8))
s.shutdown(socket.SHUT_WR)
data = b'x'
while data:
    data = s.recv(1024)
    print(data.decode(UTF8), end='')
s.close()
```

Exmaple usage
```bash
python groovy_scripting_client.py "c.statusInfo = 'Hello from Python'"
```

**Reply by macmarrum** [reply]
Save [groovyScriptingServer.groovy](https://raw.githubusercontent.com/macmarrum/freeplane-scripting/main/src/init/groovyScriptingServer.groovy) in `<user-dir>/scripts/init/` and restart Freeplane

**Comment by metacoma** [comment]
Hey @macmarrum ! 
Thank you for your interest in this topic! 

I encountered a number of limitations when using raw sockets, so I wrote a plugin that allows communication with a Freeplane instance using gRPC.

**https://github.com/metacoma/freeplane_plugin_grpc**
And topic in Freeplane discussion  **https://github.com/freeplane/freeplane/discussions/970**

The plugin is currently in the early beta stage, but it already meets my needs

**Reply by macmarrum** [reply]
> I encountered a number of limitations when using raw sockets, so I wrote a plugin that allows communication with a Freeplane instance using gRPC.

Another great idea :+1: 

Could you please elaborate on the limitations of using raw sockets?

**Reply by metacoma** [reply]
In the case of raw sockets, you would need to "invent" a client-server protocol to execute code. For example, imagine how challenging it would be to transmit a JSON/YAML/XML file through a raw socket, process it using Groovy, and render it in Freeplane.

Using raw sockets directly for such tasks can indeed be complex and error-prone

In the gRPC plugin, I have also included the [capability](https://github.com/metacoma/freeplane_plugin_grpc/blob/main/src/main/java/org/freeplane/plugin/grpc/GrpcRegistration.java#L199) to remotely execute pure Groovy code. However, I have been unable to implement it successfully so far.

**Reply by macmarrum** [reply]
Your proposed usage of gRPC is very interesting. I can see a lot of thought and energy invested in it :clap: 

> In the case of raw sockets, you would need to "invent" a client-server protocol to execute code. For example, imagine how challenging it would be to transmit a JSON/YAML/XML file through a raw socket, process it using Groovy, and render it in Freeplane.

Yes. I can see the benefits on using gRPC. At the same time I am concerned that it adds another layer of abstraction (service definitions). The way I understand it, if I wanted to add e.g. a cloud to a node, I'd have to extend the gRPC plugin for Freeplane first.

On a side note, I can see that your implementation uses internal Freeplane API, which runs the risk of breaking with any future Freeplane release. Have you considered using the [Scripting API](https://docs.freeplane.org/api/)?

**Reply by metacoma** [reply]
>At the same time I am concerned that it adds another layer of abstraction (service definitions). The way I understand it, if I wanted to add e.g. a cloud to a node, I'd have to extend the gRPC plugin for Freeplane first.

Yes, that's exactly right. To add new functionality, you need to extend the gRPC specification and write code for the function in the plugin. You can see the changes I made to implement JSON document import https://github.com/metacoma/freeplane_plugin_grpc/pull/6/files (changes were made manually to the `freeplane.proto` and `GrpcRegistration.java` files, while the rest of the changes are auto-generated code for the python/ruby libraries)

>On a side note, I can see that your implementation uses internal Freeplane API, which runs the risk of breaking with any future Freeplane release. Have you considered using the [Scripting API](https://docs.freeplane.org/api/)?

Yes, I am aware of the Scripting API. It would be great if you could provide some code examples or snippets demonstrating how I can use it in the gRPC plugin.

I have developed this plugin for my another project https://github.com/metacoma/mindwm where it is actively used, and perhaps now the plugin may not look structurally appealing. I would be happy to hear your opinion and any other suggestions on how to further develop this plugin. Ideally, I envision it being included in the main distribution of Freeplane, but for that, many improvements need to be made. I am open to ideas and would gladly listen to them.
